server/src/main/java/org/elasticsearch/index/codec/vectors/es910/ES910HnswVectorsWriter.java

```diff
@@ -0,0 +1,645 @@
+        meta.writeVLong(vectorIndexLength);
+        meta.writeVInt(field.getVectorDimension());
+        meta.writeInt(count);
+        meta.writeVInt(M);
+        // write graph nodes on each level
+        if (graph == null) {
+            meta.writeVInt(0);
+        } else {
+            meta.writeVInt(graph.numLevels());
+            long valueCount = 0;
+            for (int level = 0; level < graph.numLevels(); level++) {
+                HnswGraph.NodesIterator nodesOnLevel = graph.getNodesOnLevel(level);
+                valueCount += nodesOnLevel.size();
+                if (level > 0) {
+                    int[] nol = new int[nodesOnLevel.size()];
+                    int numberConsumed = nodesOnLevel.consume(nol);
+                    Arrays.sort(nol);
+                    assert numberConsumed == nodesOnLevel.size();
+                    meta.writeVInt(nol.length); // number of nodes on a level
+                    for (int i = nodesOnLevel.size() - 1; i > 0; --i) {
+                        nol[i] -= nol[i - 1];
+                    }
+                    for (int n : nol) {
+                        assert n >= 0 : "delta encoding for nodes failed; expected nodes to be sorted";
+                        meta.writeVInt(n);
+                    }
+                } else {
+                    assert nodesOnLevel.size() == count : "Level 0 expects to have all nodes";
+                }
+            }
+            long start = vectorIndex.getFilePointer();
+            meta.writeLong(start);
+            meta.writeVInt(DIRECT_MONOTONIC_BLOCK_SHIFT);
+            final DirectMonotonicWriter memoryOffsetsWriter = DirectMonotonicWriter.getInstance(
+                meta,
+                vectorIndex,
+                valueCount,
+                DIRECT_MONOTONIC_BLOCK_SHIFT
+            );
+            long cumulativeOffsetSum = 0;
+            for (int[] levelOffsets : graphLevelNodeOffsets) {
+                for (int v : levelOffsets) {
+                    memoryOffsetsWriter.add(cumulativeOffsetSum);
+                    cumulativeOffsetSum += v;
+                }
+            }
+            memoryOffsetsWriter.finish();
+            meta.writeLong(vectorIndex.getFilePointer() - start);
+        }
+    }
+
+    private HnswGraphMerger createGraphMerger(
+        FieldInfo fieldInfo,
+        RandomVectorScorerSupplier scorerSupplier,
+        TaskExecutor parallelMergeTaskExecutor,
+        int numParallelMergeWorkers
+    ) {
+        if (mergeExec != null) {
+            return new ConcurrentHnswMerger(fieldInfo, scorerSupplier, M, beamWidth, mergeExec, numMergeWorkers);
+        }
+        if (parallelMergeTaskExecutor != null && numParallelMergeWorkers > 1) {
+            return new ConcurrentHnswMerger(fieldInfo, scorerSupplier, M, beamWidth, parallelMergeTaskExecutor, numParallelMergeWorkers);
+        }
+        return new IncrementalHnswGraphMerger(fieldInfo, scorerSupplier, M, beamWidth);
```

Comment:
we only need `IncrementalHnswGraphMerger` we never use the task exec stuff nor do concurrent merges.